package com.pg85.otg.gen.resource;

import com.pg85.otg.constants.Constants;
import com.pg85.otg.exceptions.InvalidConfigException;
import com.pg85.otg.config.settings.biome.BiomeSettings;
import com.pg85.otg.interfaces.IWorldGenRegion;
import com.pg85.otg.util.helpers.RandomHelper;

import java.util.List;
import java.util.Random;

public final class FossilResource extends FrequencyResourceBase
{
	private final int rarity;
	private final int maxAltitude;
	private final int minAltitude;
	
	public FossilResource(BiomeSettings biomeConfig, List<String> args) throws InvalidConfigException
	{
		super(biomeConfig, args);
		assureSize(1, args);

		this.frequency = 1;
		this.rarity = readInt(args.get(0), 1, Integer.MAX_VALUE);
		if (args.size() >= 3) {
			this.minAltitude = readInt(args.get(1), Constants.WORLD_DEPTH, Constants.WORLD_HEIGHT - 1);
			this.maxAltitude = readInt(args.get(2), minAltitude, Constants.WORLD_HEIGHT - 1);
		} else {
			// Extremely rough default for legacy presets
			this.minAltitude = 30;
			this.maxAltitude = 60;
		}
	}

	@Override
	public void spawnForChunkDecoration(IWorldGenRegion worldGenRegion, Random random)
	{
		// Intentionally empty, since frequency does not work with fossils
	}

	@Override
	public String toString()
	{
		return "Fossil(" + this.rarity + "," + this.minAltitude + "," + this.maxAltitude + ")";
	}

	@Override
	public void spawn(IWorldGenRegion world, Random random, int x, int z)
	{
		int y = RandomHelper.numberInRange(random, this.minAltitude, this.maxAltitude);		
		world.placeFossil(random, x, y, z);
	}
}
